ext/drx/scatter_gather_aarch64.c

```diff
@@ -419,46 +423,54 @@ expand_scatter_gather(void *drcontext, instrlist_t *bb, instr_t *sg_instr,
     DR_ASSERT_MSG(reg_is_z(sg_info->base_reg) || reg_is_z(sg_info->index_reg),
                   "Base or index must be a Z register");
 
+    DR_ASSERT_MSG(
+        (scalar_src_or_dst != scalar_index_or_base) ||
+            (sg_info->is_load && sg_info->reg_count == 1),
+        "Single register loads are the only instructions where it is safe for the scratch"
+        " registers to alias");
+
     const uint no_of_elements =
         (opnd_size_in_bytes(sg_info->scatter_gather_size) / sg_info->reg_count) /
         opnd_size_in_bytes(sg_info->scalar_value_size);
 
-    if (sg_info->is_load) {
-        /* First we deal with the inactive elements. Gather loads are always zeroing so we
-         * need to set all inactive elements to 0.
-         */
-        if ((sg_info->base_reg == sg_info->gather_dst_reg) ||
-            (sg_info->index_reg == sg_info->gather_dst_reg)) {
-            /* The dst register is also the base/index register so we need to preserve the
-             * value of the active elements so we can use them in the address calculation.
-             * We do this by CPYing a 0 value into the dst register using the inverse of
-             * the mask_reg as the governing predicate.
-             */
-
-            /* ptrue    scratch_pred.b */
-            EMIT(ptrue_sve, opnd_create_reg_element_vector(scratch_pred, OPSZ_1),
-                 opnd_create_immed_pred_constr(DR_PRED_CONSTR_ALL));
-
-            /* not      scratch_pred.b, scratch_pred/z, mask_reg.b */
-            EMIT(not_sve_pred_b, opnd_create_reg_element_vector(scratch_pred, OPSZ_1),
-                 opnd_create_predicate_reg(scratch_pred, false),
-                 opnd_create_reg_element_vector(sg_info->mask_reg, OPSZ_1));
-
-            /* cpy      gather_dst_reg.element_size, scratch_pred/m, #0, lsl #0 */
-            EMIT(cpy_sve_shift_pred,
-                 opnd_create_reg_element_vector(sg_info->gather_dst_reg,
-                                                sg_info->element_size),
-                 opnd_create_predicate_reg(scratch_pred, true), OPND_CREATE_INT8(0),
-                 opnd_create_immed_uint(0, OPSZ_1b));
-        } else {
-            /* We don't care about any values in the dst register so zero the whole thing.
+    for (uint reg_index = 0; reg_index < sg_info->reg_count; reg_index++) {
+        const reg_id_t vector_dst =
+            (((sg_info->gather_dst_reg - DR_REG_Z0) + reg_index) % 32) + DR_REG_Z0;
+        if (sg_info->is_load) {
```

Comment:
Looks like we can avoid the loop completely for !is_load if we move the condition before the loop.

Reply:
I just spotted that too. Done.